### PR TITLE
[MM-27556] UI For new system roles

### PR DESCRIPTION
**Pull request by fmunshi**
#### Summary
- Adds UI to assign and edit system roles

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-27556

#### Related Pull Requests
- Has server changes https://github.com/mattermost/mattermost-server/pull/15500
- Has redux changes https://github.com/mattermost/mattermost-redux/pull/1269

#### Screenshots
Screen Shot 2020-09-22 at 4 16 17 PM
Screen Shot 2020-09-22 at 4 15 52 PM

### CODE REVIEW

File: components/admin_console/data_grid/data_grid.tsx [hunk 1]
```diff
@@ -52,8 +52,8 @@ type Props = {
     nextPage: () => void;
     previousPage: () => void;
 
-    search: (term: string) => void;
-    term: string;
+    search?: (term: string) => void;
```
Comment (hahmadia) [inline components/admin_console/data_grid/data_grid.tsx:55]: Small realization put would renaming this to `onSearch` require too many changes? Just feel it suites better but non-blocking cause I would think many other places would require the change.

File: components/admin_console/system_roles/system_role/system_role_users/system_role_users.tsx [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+            getFilteredUsersStats({roles: [this.props.role.name]}),
+        ]);
+        this.setStateLoading(false);
+    }
+
+    async componentDidUpdate(prevProps: Props) {
+        if (prevProps.term !== this.props.term) {
+            this.setStateLoading(true);
+            clearTimeout(this.searchTimeoutId);
+            const term = this.props.term;
```
Comment (hahmadia) [inline components/admin_console/system_roles/system_role/system_role_users/system_role_users.tsx:84]: maybe better to do {term} = this.props

File: components/admin_console/system_roles/system_role/system_role_users/system_role_users.tsx [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        const {term, role, usersToAdd, usersToRemove, readOnly} = this.props;
+        const {startCount, endCount, total} = this.getPaginationProps();
+        return (
+
+            <AdminPanel
+                id='SystemRoleUsers'
+                titleId={t('admin.permissions.system_role_users.title')}
+                titleDefault='Assigned People'
+                subtitleId={t('admin.permissions.system_role_users.description')}
+                subtitleDefault={'List of people assigned to this system role.'}
```
Comment (hahmadia) [inline components/admin_console/system_roles/system_role/system_role_users/system_role_users.tsx:256]:
```suggestion
                subtitleDefault='List of people assigned to this system role.'
```

**Review by hahmadia — COMMENTED** [review]
Some small changes. Also, does it make sense to do UX review on this or nah? I wasn't sure so I thought i ask

File: components/admin_console/data_grid/data_grid.tsx (hunk 1 above)
Comment (fmunshi) [reply components/admin_console/data_grid/data_grid.tsx:55]: yea I agree, will make the change

File: components/admin_console/system_roles/index.tsx [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+import {connect} from 'react-redux';
+
+import {getRoles} from 'mattermost-redux/selectors/entities/roles_helpers';
+
+import {GlobalState} from 'types/store';
+
+import SystemRoles from './system_roles';
+
+function mapStateToProps(state: GlobalState) {
+    const roles = (({system_admin, system_user_manager, system_manager, system_read_only_admin}) => ({system_admin, system_user_manager, system_manager, system_read_only_admin}))(getRoles(state)); // eslint-disable-line @typescript-eslint/naming-convention
```
[the remark this answers is not in the document]
Comment (fmunshi) [reply components/admin_console/system_roles/index.tsx:13]: Its basically an anonymous function being called with the only parameter as the return value of `getRoles(state)`.

the roles map is then being de-structured in the params and returning an object made up of the de-structured objects...

Its basically the equivalent of doing 
```
const allRoles = getRoles(state);
const {system_admin, system_user_manager, system_manager, system_read_only_admin} = allRoles;
const roles = {
  system_admin,
  system_user_manager,
  system_manager,
  system_read_only_admin
};
```

**Comment by fmunshi** [comment]
UX review will probably just be CSS changes... but I dont think i'll be around for that either you (@hahmadia) or @mkraft will have to co-ordinate with @michaelgamble to get that done

**Comment by jasonblais** [comment]
@mkraft @hahmadia will this be picked up by someone from your team?

**Comment by hahmadia** [comment]
@jasonblais Yep! Our team is looking at this.

**Review by mkraft — APPROVED** [review]

**Review by hahmadia — APPROVED** [review]

**Review by srkgupta — CHANGES_REQUESTED** [review]
Hi @hahmadia 

I am seeing the following issues (some of which is preventing me to test further). Can you please fix the same and re-request the review so that this can be tested further. 

[MM-29891](https://mattermost.atlassian.net/browse/MM-29891) - System Console > System Roles page should only be displayed on E20 instance and should not be displayed on E10 instance

[MM-29892](https://mattermost.atlassian.net/browse/MM-29892) - No option found to add additional sections for System Roles 

[MM-29894](https://mattermost.atlassian.net/browse/MM-29894)  - JS console error prevents Sysadmin to customize and add additional permissions to System Roles

[MM-29895](https://mattermost.atlassian.net/browse/MM-29895) - System Roles - Unable to modify Read Only role by sysadmin

[MM-29896](https://mattermost.atlassian.net/browse/MM-29896) - System Roles > Add Users dialog header is not displayed correctly

**Comment by srkgupta** [comment]
Also one more thing @hahmadia, can you please capitalize the word "Mattermost" in all the help text:
Screenshot 2020-10-21 at 7 13 16 PM

**Comment by hahmadia** [comment]
@srkgupta 
Ok, I think I've addressed all with exception of the ones we spoke about in the DM's.
You will need to run these two commands while I Figure out how to migrate these two permissions but in the meantime you can run this and things should work fine
```
➜  ~ mmctl permissions add system_admin sysconsole_read_user_management_system_roles --local
➜  ~ mmctl permissions add system_admin sysconsole_write_user_management_system_roles --local
```

**Comment by hahmadia** [comment]
@srkgupta Also, note that there is redux changes you will need to pull and use.

**Review by srkgupta — CHANGES_REQUESTED** [review]
Hi @hahmadia 

The System Console is not loading up with your latest changes. I am seeing the following error: 

Screenshot 2020-10-22 at 6 42 33 PM

This issue is also seen on the Cloud Test instance created. Can you please check and fix this issue ASAP.

**Review by srkgupta — CHANGES_REQUESTED** [review]
Thanks @hahmadia. With the latest changes, the previous important issues are fixed now and I am now able to proceed further with my testing. 

I am seeing the following new issues. Can you please fix them:

MM-29926 - System Roles permission should always be no_access and should not be allowed to be assigned to other system admin roles

MM-29927 - Discard Changes confirmation is not displayed when user makes some changes in System Roles page and clicks somewhere else

As per discussion with @thefactremains, following issues will be addressed later:

https://mattermost.atlassian.net/browse/MM-29892
https://mattermost.atlassian.net/browse/MM-29895

**Comment by hahmadia** [comment]
@srkgupta 

https://mattermost.atlassian.net/browse/MM-29895 -> Read Only role is now editable(with exception to system roles ... Only adming has this, nobody else has it and nobody else can have it for now). This came from discussions we had with team.

https://mattermost.atlassian.net/browse/MM-29926 -> System Roles is always no access for everyone except system admin. Nobody can get it either. This came from discussions with team.

**Review by sbishel — APPROVED** [review]
I got nothin', LGTM

**Review by mkraft — APPROVED** [review]

**Review by furqanmlk — APPROVED** [review]
Tested all test cases 
https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin%3Acom.kanoah.test-manager__main-project-page#!/design?projectId=10302

Verified following issues
MM-29892 , MM-29895 , MM-29926 , MM-29927

**Review by srkgupta — APPROVED** [review]
Tested by Furkan Malik. Approving the PR

**Pull request merged**